 Introduction to Pandas

In [9]:
import pandas as pd
import numpy as np

Understanding Series & DataFrames

Series represents one column
Combine multiple columns to create a table ( .i.e DataFrame )

In [10]:
ser1 = pd.Series(data=[1,2,3,4,5], index=list('abcde'))
ser1

a    1
b    2
c    3
d    4
e    5
dtype: int64

In [11]:
ser2 = pd.Series(data=[11,22,33,44,55], index=list('abcde'))
ser2

a    11
b    22
c    33
d    44
e    55
dtype: int64

Creating DataFrame from above two series
Data corresponding to same index belongs to same row

In [12]:
df = pd.DataFrame({'A':ser1, 'B':ser2})
df

A   B
a  1  11
b  2  22
c  3  33
d  4  44
e  5  55

Creating a random dataframe of 10 X 10

In [13]:
pd.DataFrame(data=np.random.randint(1,10,size=(10,10)), index=list('ABCDEFGHIJ'), columns=list('abcdefghij'))

a  b  c  d  e  f  g  h  i  j
A  2  8  4  4  2  8  9  5  8  1
B  5  4  8  5  8  2  9  5  8  6
C  5  8  4  8  5  8  6  5  6  9
D  6  7  1  7  3  1  8  6  9  2
E  8  8  3  4  5  3  8  2  7  9
F  6  2  9  3  9  2  5  7  3  9
G  7  9  7  7  3  9  2  4  1  7
H  7  8  6  7  9  2  8  2  3  2
I  6  8  1  4  7  4  2  2  5  4
J  3  2  5  7  9  2  6  5  7  7

Loading CSV,JSON

In [14]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [15]:
hr_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [16]:
hr_data_itr = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt', chunksize=5000)

In [17]:
for hr_data in hr_data_itr:
    print (hr_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     5000 non-null   float64
 1   last_evaluation        5000 non-null   float64
 2   number_project         5000 non-null   int64  
 3   average_montly_hours   5000 non-null   int64  
 4   time_spend_company     5000 non-null   int64  
 5   Work_accident          5000 non-null   int64  
 6   left                   5000 non-null   int64  
 7   promotion_last_5years  5000 non-null   int64  
 8   sales                  5000 non-null   object 
 9   salary                 5000 non-null   object 
dtypes: float64(2), int64(6), object(2)
memory usage: 390.8+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 5000 to 9999
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 ----------

In [18]:
pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

Descriptive Statistics

In [19]:
hr_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/HR_comma_sep.csv.txt')

In [20]:
hr_data.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years  sales  \
0                   3              0     1                      0  sales   
1                   6              0     1                      0  sales   
2                   4              0     1                      0  sales   
3                   5              0     1                      0  sales   
4                   3              0     1                      0  sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [21]:
hr_data.tail()

satisfaction_level  last_evaluation  number_project  \
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years    sales salary  
14994                      0  support    low  
14995                      0  support    low  
14996                      0  support    low  
14997                      0  support    low  
14998                      0  support    low

In [22]:
hr_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   sales                  14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [23]:
hr_data.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [24]:
hr_data.salary.value_counts()

low       7316
medium    6446
high      1237
Name: salary, dtype: int64

Accessing subset of data - rows, columns, filters

Get all columns with categorical values

In [25]:
cat_cols_data = hr_data.select_dtypes('object')

In [26]:
cat_cols_data.head()

sales  salary
0  sales     low
1  sales  medium
2  sales  medium
3  sales     low
4  sales     low

Rename columns names

In [27]:
hr_data.rename(columns={'sales':'department'},inplace=True)

In [28]:
hr_data.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

Select column by column names

In [29]:
hr_data.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'department', 'salary'],
      dtype='object')

In [30]:
hr_data[['satisfaction_level','last_evaluation','number_project']].head()

satisfaction_level  last_evaluation  number_project
0                0.38             0.53               2
1                0.80             0.86               5
2                0.11             0.88               7
3                0.72             0.87               5
4                0.37             0.52               2

In [31]:
hr_data.satisfaction_level[:5]

0    0.38
1    0.80
2    0.11
3    0.72
4    0.37
Name: satisfaction_level, dtype: float64

In [32]:
movie_data = pd.read_json('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/movie.json.txt')

In [33]:
movie_data

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

In [34]:
movie_data.loc['Scarface']

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [36]:
movie_data.loc['Scarface':'Roman Holiday']

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

In [37]:
movie_data.iloc[1]

David Smith         4.5
Brenda Peterson     1.5
Bill Duffy          5.0
Samuel Miller       3.5
Julie Hammel        2.5
Clarissa Jackson    4.5
Adam Cohen          3.0
Chris Duncan        NaN
Name: Scarface, dtype: float64

In [38]:
movie_data.iloc[1:4]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Scarface             4.5              1.5         5.0            3.5   
Raging Bull          3.0              1.0         NaN            5.0   
Goodfellas           4.5              2.0         4.5            5.0   

             Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Scarface              2.5               4.5         3.0           NaN  
Raging Bull           NaN               4.0         NaN           4.5  
Goodfellas            3.0               2.5         4.5           NaN

Filtering rows based on conditions

In [39]:
movie_data[ (movie_data['Adam Cohen'] > 3)]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo             4.0              3.0         4.5            NaN   
Goodfellas          4.5              2.0         4.5            5.0   

            Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo              NaN               5.0         3.5           NaN  
Goodfellas           3.0               2.5         4.5           NaN

In [40]:
movie_data[ ((movie_data['Adam Cohen'] > 3) & (movie_data['David Smith'] > 4))]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Goodfellas          4.5              2.0         4.5            5.0   

            Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Goodfellas           3.0               2.5         4.5           NaN

Handling missing data

Machine Learning algorithms don't expect data missing
If there is a columns with more than 40% data missing, we may drop the column
Fow rows with, important column values missing, drop the rows

In [41]:
movie_data

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Raging Bull            3.0              1.0         NaN            5.0   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Raging Bull             NaN               4.0         NaN           4.5  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5  
Roman Holiday           4.5               1.5         3.0           NaN

In [42]:
movie_data['Bill Duffy'].notnull()

Vertigo           True
Scarface          True
Raging Bull      False
Goodfellas        True
The Apartment     True
Roman Holiday    False
Name: Bill Duffy, dtype: bool

In [43]:
movie_data[movie_data['Bill Duffy'].notnull()]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Vertigo                4.0              3.0         4.5            NaN   
Scarface               4.5              1.5         5.0            3.5   
Goodfellas             4.5              2.0         4.5            5.0   
The Apartment          1.0              5.0         1.0            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Vertigo                 NaN               5.0         3.5           NaN  
Scarface                2.5               4.5         3.0           NaN  
Goodfellas              3.0               2.5         4.5           NaN  
The Apartment           NaN               1.0         1.0           1.5

Get all the rows for which 'Bill Duffy' is null

In [44]:
movie_data[movie_data['Bill Duffy'].isnull()]

David Smith  Brenda Peterson  Bill Duffy  Samuel Miller  \
Raging Bull            3.0              1.0         NaN            5.0   
Roman Holiday          NaN              4.5         NaN            1.0   

               Julie Hammel  Clarissa Jackson  Adam Cohen  Chris Duncan  
Raging Bull             NaN               4.0         NaN           4.5  
Roman Holiday           4.5               1.5         3.0           NaN

Dropping Rows & Columns

In [45]:
titanic_data = pd.read_csv('https://raw.githubusercontent.com/zekelabs/data-science-complete-tutorial/master/Data/titanic-train.csv.txt')

In [46]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Dropping 'Cabin' column as it has only 204 data present in 891 rows

In [47]:
titanic_data.drop(['Cabin'],axis=1,inplace=True)

In [48]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


Now, drop all rows with missing values
We don't have inplace = True, so doesn't modify the dataframe

In [49]:
titanic_data.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


Consider only selected columns to check if they contain NA

In [50]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [51]:
titanic_data.dropna(subset=['Embarked','Age']).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  712 non-null    int64  
 1   Survived     712 non-null    int64  
 2   Pclass       712 non-null    int64  
 3   Name         712 non-null    object 
 4   Sex          712 non-null    object 
 5   Age          712 non-null    float64
 6   SibSp        712 non-null    int64  
 7   Parch        712 non-null    int64  
 8   Ticket       712 non-null    object 
 9   Fare         712 non-null    float64
 10  Embarked     712 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 66.8+ KB


Another approach of handling missing data is filling the missing ones

In [52]:
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [53]:
titanic_data.fillna({'Age':0,'Embarked':'Unknown'}).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 76.7+ KB


In [54]:
titanic_data.Age.fillna(method='ffill')[:5]
#Other options are 'bfill'

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

Handling Duplicates

In [56]:
df = pd.DataFrame({'A':[1,1,3,4,5,1], 'B':[1,1,3,7,8,1], 'C':[3,1,1,6,7,1]})
df

A  B  C
0  1  1  3
1  1  1  1
2  3  3  1
3  4  7  6
4  5  8  7
5  1  1  1

In [57]:
df.duplicated()

0    False
1    False
2    False
3    False
4    False
5     True
dtype: bool

In [58]:
df[df.duplicated()]

A  B  C
5  1  1  1

In [59]:
df[df.duplicated(subset=['A','B'])]

A  B  C
1  1  1  1
5  1  1  1

Function Application

map for transforming one column to another
Can be applied only to series

In [60]:
titanic_data_age = titanic_data[titanic_data.Age.notnull()]

In [61]:
titanic_data['age_category'] = titanic_data.Age.map(lambda age: 'Kid' if age < 18 else 'Adult')

In [62]:
titanic_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked age_category  
0      0         A/5 21171   7.2500        S        Adult  
1      0          PC 17599  71.2833        C        Adult  
2      0  STON/O2. 3101282   7.9250        S        Adult  
3      0            113803  53.1000        S        Adult  
4      0            373450   8.0500        S        Adult

apply function can be done to Series as well as DataFrames

In [63]:
titanic_data.Age.apply('sum')

21205.17

In [64]:
titanic_data.Age.apply(lambda age: 'Kid' if age < 18 else 'Adult')[:10]

0    Adult
1    Adult
2    Adult
3    Adult
4    Adult
5    Adult
6    Adult
7      Kid
8    Adult
9      Kid
Name: Age, dtype: object

apply on dataframes helps us dealing with multiple columns
func will receive all the rows

In [65]:
#e will be each row
def func(e):
    if e.Sex == 'male':
        return e.Fare * 2
    else:
        return e.Fare

In [66]:
titanic_data.apply(func,axis=1)[:5]

0    14.5000
1    71.2833
2     7.9250
3    53.1000
4    16.1000
dtype: float64

groupby - It splits data into groups, a function is applied to each groups separately, combine results into a data structure

In [67]:
titanic_data.groupby(['Sex']).Age.mean()

Sex
female    27.915709
male      30.726645
Name: Age, dtype: float64

In [68]:
titanic_data.groupby(['Sex']).Age.agg(['mean','min','max'])

mean   min   max
Sex                          
female  27.915709  0.75  63.0
male    30.726645  0.42  80.0

Rolling for window based operation

In [69]:
titanic_data.Age.rolling(window=5,min_periods=1).agg(['sum','min'])

sum   min
0     22.0  22.0
1     60.0  22.0
2     86.0  22.0
3    121.0  22.0
4    156.0  22.0
..     ...   ...
886  141.0  22.0
887  138.0  19.0
888  110.0  19.0
889  111.0  19.0
890  104.0  19.0

[891 rows x 2 columns]

In [70]:
titanic_data[titanic_data.Name.str.contains('Mr')]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
3              4         1       1   
4              5         0       3   
5              6         0       3   
..           ...       ...     ...   
883          884         0       2   
884          885         0       3   
885          886         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
883                      Banfield, Mr. Frederick James    male  28.0      0   
884                             Sutehall, Mr. Henry Jr    male  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Embarked age_category  
0        0         A/5 21171   7.2500        S        Adult  
1        0          PC 17599  71.2833        C        Adult  
3        0            113803  53.1000        S        Adult  
4        0            373450   8.0500        S        Adult  
5        0            330877   8.4583        Q        Adult  
..     ...               ...      ...      ...          ...  
883      0  C.A./SOTON 34068  10.5000        S        Adult  
884      0   SOTON/OQ 392076   7.0500        S        Adult  
885      5            382652  29.1250        Q        Adult  
889      0            111369  30.0000        C        Adult  
890      0            370376   7.7500        Q        Adult  

[647 rows x 12 columns]

Append,Merge, Join & Concatenate

Append for stacking dataframe

In [73]:
df1 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))
df1

A  B  C
0  6  2  8
1  5  6  6
2  8  3  7
3  2  6  8
4  1  1  3
5  2  7  3
6  1  7  4
7  8  8  8
8  5  8  9
9  8  4  9

In [74]:
df2 = pd.DataFrame(data=np.random.randint(1,10,size=(10,3)), columns=list('ABC'))
df2

A  B  C
0  3  3  6
1  1  4  5
2  8  8  8
3  4  7  4
4  5  3  2
5  6  4  2
6  6  6  7
7  1  4  1
8  7  4  5
9  4  9  7

In [75]:
df1.append(df2, ignore_index=True)

A  B  C
0   6  2  8
1   5  6  6
2   8  3  7
3   2  6  8
4   1  1  3
5   2  7  3
6   1  7  4
7   8  8  8
8   5  8  9
9   8  4  9
10  3  3  6
11  1  4  5
12  8  8  8
13  4  7  4
14  5  3  2
15  6  4  2
16  6  6  7
17  1  4  1
18  7  4  5
19  4  9  7

In [76]:
left = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K4','K5'],
                         'A': ['A0', 'A1', 'A2', 'A3','A4','A5'],
                         'B': ['B0', 'B1', 'B2', 'B3','B4','B5']})

right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3','K6','K7'],
                          'C': ['C0', 'C1', 'C2', 'C3','C6','C7'],
                         'D': ['D0', 'D1', 'D2', 'D3','D6','D7']})

In [77]:
left.merge(right, on='key')

key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K1  A1  B1  C1  D1
2  K2  A2  B2  C2  D2
3  K3  A3  B3  C3  D3

In [78]:
left.merge(right, on='key', how='left')

key   A   B    C    D
0  K0  A0  B0   C0   D0
1  K1  A1  B1   C1   D1
2  K2  A2  B2   C2   D2
3  K3  A3  B3   C3   D3
4  K4  A4  B4  NaN  NaN
5  K5  A5  B5  NaN  NaN

join for combining data based on index values

In [79]:
left = pd.DataFrame({'A': ['A0', 'A1', 'A2'],
                     'B': ['B0', 'B1', 'B2']},
                     index=['K0', 'K1', 'K2'])

right = pd.DataFrame({'C': ['C0', 'C2', 'C3'],
                      'D': ['D0', 'D2', 'D3']},
                     index=['K0', 'K2', 'K3'])

In [80]:
left.join(right)

A   B    C    D
K0  A0  B0   C0   D0
K1  A1  B1  NaN  NaN
K2  A2  B2   C2   D2

Pivot Tables

In [ ]:
A useful way to get important information from data

In [81]:
sales_data = pd.read_excel('https://github.com/zekelabs/data-science-complete-tutorial/blob/master/Data/sales-funnel.xlsx?raw=true')

In [82]:
sales_data

Account                          Name            Rep        Manager  \
0    714466               Trantow-Barrows   Craig Booker   Debra Henley   
1    714466               Trantow-Barrows   Craig Booker   Debra Henley   
2    714466               Trantow-Barrows   Craig Booker   Debra Henley   
3    737550  Fritsch, Russel and Anderson   Craig Booker   Debra Henley   
4    146832                  Kiehn-Spinka  Daniel Hilton   Debra Henley   
5    218895                     Kulas Inc  Daniel Hilton   Debra Henley   
6    218895                     Kulas Inc  Daniel Hilton   Debra Henley   
7    412290                 Jerde-Hilpert     John Smith   Debra Henley   
8    740150                    Barton LLC     John Smith   Debra Henley   
9    141962                    Herman LLC    Cedric Moss  Fred Anderson   
10   163416                   Purdy-Kunde    Cedric Moss  Fred Anderson   
11   239344                    Stokes LLC    Cedric Moss  Fred Anderson   
12   239344                    Stokes LLC    Cedric Moss  Fred Anderson   
13   307599   Kassulke, Ondricka and Metz     Wendy Yule  Fred Anderson   
14   688981                   Keeling LLC     Wendy Yule  Fred Anderson   
15   729833                     Koepp Ltd     Wendy Yule  Fred Anderson   
16   729833                     Koepp Ltd     Wendy Yule  Fred Anderson   

        Product  Quantity   Price     Status  
0           CPU         1   30000  presented  
1      Software         1   10000  presented  
2   Maintenance         2    5000    pending  
3           CPU         1   35000   declined  
4           CPU         2   65000        won  
5           CPU         2   40000    pending  
6      Software         1   10000  presented  
7   Maintenance         2    5000    pending  
8           CPU         1   35000   declined  
9           CPU         2   65000        won  
10          CPU         1   30000  presented  
11  Maintenance         1    5000    pending  
12     Software         1   10000  presented  
13  Maintenance         3    7000        won  
14          CPU         5  100000        won  
15          CPU         2   65000   declined  
16      Monitor         2    5000  presented

In [83]:
pd.pivot_table(sales_data, index=['Manager','Rep'], values=['Account','Price'], aggfunc=[np.sum, np.mean])

sum              mean              
                             Account   Price   Account         Price
Manager       Rep                                                   
Debra Henley  Craig Booker   2880948   80000  720237.0  20000.000000
              Daniel Hilton   584622  115000  194874.0  38333.333333
              John Smith     1152440   40000  576220.0  20000.000000
Fred Anderson Cedric Moss     784066  110000  196016.5  27500.000000
              Wendy Yule     2456246  177000  614061.5  44250.000000

Normalizing JSON